src/ansi-c/library/cprover_contracts.c

```diff
@@ -0,0 +1,4116 @@
+  // (replacement mode only)
+  __CPROVER_contracts_obj_set_ptr_t linked_allocated;
+  // true iff this write set is used for contract replacement
+  __CPROVER_bool replacement;
+  // true iff this write set checks requires clauses in an assumption ctx
+  __CPROVER_bool assume_requires_ctx;
+  // true iff this write set checks requires clauses in an assertion ctx
+  __CPROVER_bool assert_requires_ctx;
+  // true iff this write set checks ensures clauses in an assumption ctx
+  __CPROVER_bool assume_ensures_ctx;
+  // true iff this write set checks ensures clauses in an assertion ctx
+  __CPROVER_bool assert_ensures_ctx;
+} __CPROVER_contracts_write_set_t;
+
+typedef __CPROVER_contracts_write_set_t *__CPROVER_contracts_write_set_ptr_t;
+#endif
+
+inline __CPROVER_contracts_car_t
+__CPROVER_contracts_car_create(void *ptr, __CPROVER_size_t size)
+{
+#pragma CPROVER check push
+#pragma CPROVER check disable "pointer"
+#pragma CPROVER check disable "pointer-primitive"
+#pragma CPROVER check disable "unsigned-overflow"
+#pragma CPROVER check disable "signed-overflow"
+#pragma CPROVER check disable "undefined-shift"
+#pragma CPROVER check disable "conversion"
+  __CPROVER_assert(
+    ((ptr == 0) | __CPROVER_rw_ok(ptr, size)),
+    "ptr NULL or writable up to size");
+  __CPROVER_contracts_car_t car = {0};
+  car.is_writable = ptr != 0;
+  car.size = size;
+  car.lb = ptr;
+  car.ub = (void *)((char *)ptr + size);
+  __CPROVER_assert(
+    __CPROVER_same_object(car.lb, car.ub), "no offset bits overflow on ub");
+#pragma CPROVER check pop
+  return car;
+}
+
+/* FUNCTION: __CPROVER_contracts_car_set_create */
+
+#ifndef __CPROVER_contracts_write_set_t_defined
+#  define __CPROVER_contracts_write_set_t_defined
+
+typedef struct __CPROVER_contracts_car_t
+{
+  __CPROVER_bool is_writable;
+  __CPROVER_size_t size;
+  void *lb;
+  void *ub;
+} __CPROVER_contracts_car_t;
+
+typedef struct __CPROVER_contracts_car_set_t
+{
+  // maximum number of elements that can be stored in the set
+  __CPROVER_size_t max_elems;
+  // 1 + greatest used index in elems,
+  // next usable index in elems if less than max_elems
+  __CPROVER_size_t watermark;
+  // A array of car_t of size max_elems
+  __CPROVER_contracts_car_t *elems;
+} __CPROVER_contracts_car_set_t;
+
+typedef __CPROVER_contracts_car_set_t *__CPROVER_contracts_car_set_ptr_t;
+
+typedef struct __CPROVER_contracts_obj_set_t
+{
+  // maximum number of elements that can be stored in the set
+  __CPROVER_size_t max_elems;
+  // 1 + greatest used index in elems,
+  // next usable index in elems if less than max_elems
+  __CPROVER_size_t watermark;
+  // number of elements currently in the elems array
+  __CPROVER_size_t nof_elems;
+  // true iff nof_elems is 0
+  __CPROVER_bool is_empty;
+  // true iff elems is indexed by the object id of the pointers
+  __CPROVER_bool indexed_by_object_id;
+  // array of void *pointers, indexed by their object ID
+  // or some other order
+  void **elems;
```

Comment:
This is kind of a weird structure when "indexed_by_object_id" is true, right? Each element should be either NULL or a pointer who's object bits are the same as the index and who's lower bits are 0.  We could probably do this with an array of booleans instead. Right? (But then you would have a different shape for when "indexed_by_object_id" is false so maybe this is better?)